### PR TITLE
[REVIEW] Add `is_index_type` trait

### LINKED ISSUE #5432: [FEA] Create a libcudf is_index_type trait
Reference this comment: https://github.com/rapidsai/cudf/pull/5410#pullrequestreview-427317812

I found the following places that use integral-but-not-bool `std::enable_if` statements and could benefit from an `cudf::is_index_type` trait

- https://github.com/rapidsai/cudf/blob/2db074454d7919693623716e2d617dabc2494f4f/cpp/src/copying/gather.cu#L21-L22
- https://github.com/rapidsai/cudf/blob/2db074454d7919693623716e2d617dabc2494f4f/cpp/src/copying/scatter.cu#L43-L44
- https://github.com/rapidsai/cudf/blob/2db074454d7919693623716e2d617dabc2494f4f/cpp/src/partitioning/partitioning.cu#L645
- https://github.com/rapidsai/cudf/blob/2db074454d7919693623716e2d617dabc2494f4f/cpp/src/filling/sequence.cu#L58

There may be more as evidence by PR #5410

### PR DESCRIPTION
**Pull request by rgsl888prabhu**
closes #5432

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

File: cpp/include/cudf/utilities/traits.hpp [hunk 1]
```diff
@@ -171,6 +171,21 @@ constexpr inline bool is_numeric(data_type type)
...
+ * @brief Indicates whether the type `T` is index type.
+ *
+ * index types are integral types such as 'INT*' apart from 'bool'.
+ *
+ * @tparam T  The type to verify
+ * @return true `T` is index type
+ * @return false  `T` is not index type
+ **/
+template <typename T>
+constexpr inline bool is_index_type()
```
Comment (jrhemstad) [inline cpp/include/cudf/utilities/traits.hpp:184]: Add one that works for `data_type` too.

File: cpp/include/cudf/utilities/traits.hpp [hunk 2]
```diff
@@ -171,6 +171,21 @@ constexpr inline bool is_numeric(data_type type)
   return cudf::type_dispatcher(type, is_numeric_impl{});
 }
 
+/**
+ * @brief Indicates whether the type `T` is index type.
+ *
+ * index types are integral types such as 'INT*' apart from 'bool'.
```
Comment (jrhemstad) [inline cpp/include/cudf/utilities/traits.hpp:177]:
```suggestion
 * A type `T` is considered an index type if it is valid to use elements of type `T` to index into column. I.e., 
 * index types are integral types such as 'INT*' apart from 'bool'.
```

Comment (harrism) [reply cpp/include/cudf/utilities/traits.hpp:177]:
```suggestion
 * A type `T` is considered an index type if it is valid to use elements of type `T` to index into a column.
 * In other words, index types are integral types such as 'INT*' apart from 'bool'.
```

File: CHANGELOG.md [hunk 3]
```diff
@@ -65,6 +65,7 @@
 - PR #5437 Improve libcudf join documentation
 - PR #5458 Install meta packages for dependencies
 - PR #5467 Move doc customization scripts to Jenkins
+- PR #5477 Add `index_type` trait
```
Comment (harrism) [inline CHANGELOG.md:68]:
```suggestion
- PR #5477 Add `is_index_type` trait 
```

**Pull request merged**